File: generated-code/ruby/codecraft/TcpReadWrite/model/entity_action.rb
require './model/attack_action'
require './model/build_action'
require './model/move_action'
require './model/repair_action'

module Model

# Entity's action
class EntityAction
    # Move action
    attr_accessor :move_action
    # Build action
    attr_accessor :build_action
    # Attack action
    attr_accessor :attack_action
    # Repair action
    attr_accessor :repair_action

    def initialize(move_action, build_action, attack_action, repair_action)
        @move_action = move_action
        @build_action = build_action
        @attack_action = attack_action
        @repair_action = repair_action
    end

    # Read EntityAction from input stream
    def self.read_from(stream)
        if stream.read_bool()
            move_action = Model::MoveAction.read_from(stream)
        else
            move_action = nil
        end
        if stream.read_bool()
            build_action = Model::BuildAction.read_from(stream)
        else
            build_action = nil
        end
        if stream.read_bool()
            attack_action = Model::AttackAction.read_from(stream)
        else
            attack_action = nil
        end
        if stream.read_bool()
            repair_action = Model::RepairAction.read_from(stream)
        else
            repair_action = nil
        end
        EntityAction.new(move_action, build_action, attack_action, repair_action)
    end

    # Write EntityAction to output stream
    def write_to(stream)
        if @move_action.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @move_action.write_to(stream)
        end
        if @build_action.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @build_action.write_to(stream)
        end
        if @attack_action.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @attack_action.write_to(stream)
        end
        if @repair_action.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @repair_action.write_to(stream)
        end
    end

    def to_s
        string_result = "EntityAction { "
        string_result += "move_action: "
        if @move_action.nil?
            string_result += "nil"
        else
            string_result += @move_action.to_s
        end
        string_result += ", "
        string_result += "build_action: "
        if @build_action.nil?
            string_result += "nil"
        else
            string_result += @build_action.to_s
        end
        string_result += ", "
        string_result += "attack_action: "
        if @attack_action.nil?
            string_result += "nil"
        else
            string_result += @attack_action.to_s
        end
        string_result += ", "
        string_result += "repair_action: "
        if @repair_action.nil?
            string_result += "nil"
        else
            string_result += @repair_action.to_s
        end
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end

File: generated-code/python/codecraft/TcpReadWrite/model/entity_action.py
from model.attack_action import AttackAction
from model.build_action import BuildAction
from model.move_action import MoveAction
from model.repair_action import RepairAction
from stream_wrapper import StreamWrapper
from typing import Optional

class EntityAction:
    """Entity's action"""

    __slots__ = ("move_action","build_action","attack_action","repair_action",)

    move_action: Optional[MoveAction]
    build_action: Optional[BuildAction]
    attack_action: Optional[AttackAction]
    repair_action: Optional[RepairAction]

    def __init__(self, move_action: Optional[MoveAction], build_action: Optional[BuildAction], attack_action: Optional[AttackAction], repair_action: Optional[RepairAction]):
        self.move_action = move_action
        """Move action"""
        self.build_action = build_action
        """Build action"""
        self.attack_action = attack_action
        """Attack action"""
        self.repair_action = repair_action
        """Repair action"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "EntityAction":
        """Read EntityAction from input stream
        """
        if stream.read_bool():
            move_action = MoveAction.read_from(stream)
        else:
            move_action = None
        if stream.read_bool():
            build_action = BuildAction.read_from(stream)
        else:
            build_action = None
        if stream.read_bool():
            attack_action = AttackAction.read_from(stream)
        else:
            attack_action = None
        if stream.read_bool():
            repair_action = RepairAction.read_from(stream)
        else:
            repair_action = None
        return EntityAction(move_action, build_action, attack_action, repair_action)
    
    def write_to(self, stream: StreamWrapper):
        """Write EntityAction to output stream
        """
        if self.move_action is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.move_action.write_to(stream)
        if self.build_action is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.build_action.write_to(stream)
        if self.attack_action is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.attack_action.write_to(stream)
        if self.repair_action is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.repair_action.write_to(stream)
    
    def __repr__(self):
        return "EntityAction(" + \
            repr(self.move_action) + \
            ", " + \
            repr(self.build_action) + \
            ", " + \
            repr(self.attack_action) + \
            ", " + \
            repr(self.repair_action) + \
            ")"